### ISSUE TITLE
data storage requirements

### ISSUE THREAD

**Issue by balmas** [alpheios-project/alignment-editor-new#19]
1. If a user has not logged in, they will not be able to save and retrieve their alignments to/from an Alpheios remote data store. (They should still be able to download them to their local machine and reupload them to work on them some more).    

2. If a user has logged in, they must be able to save and retrieve their alignments to/from an Alpheios remote data store (e.g. so that they can return to an incomplete alignment for continued work, add additional translation texts, etc).   

3. For alignments saved in the Alpheios data store, the owner of that alignment must be able to specify 
    1. access level of private of public
        1. private access means only someone logged in with the creator's userid can view the data
        2. public access means that the data in the alignment is available for read-only viewing by others and will be included in search results
    1. which information about your identity is visible with the alignment** 

4. For alignments saved in the  Alpheios data store, users should be able to set a default level of visibility of the data (per previous requirement) and be able to override that per alignment

** the available user identification information will depends upon Alpheios user account features, additional requirements for which will be tracked separately from this issue.

**Comment by balmas** [comment]
this has a dependency on the (yet to be built) Alpheios Data Service which will provide an API for saving and retrieving alignment data.

**Comment by balmas** [comment]
@irina060981  @monzug @abrasax I would like to propose these as the user data stories in scope for the first release of the alignment editor, based upon the requirements as stated in this issue.  Please let me know your thoughts.

[UDS 1] When working on an alignment, a user should have their work persist locally across browser sessions.
[UDS 2] When working on an alignment, a user should have their work persist locally while offline.
[UDS 3] When working on an large alignment with many translations, a user should be able to load only a portion of the data into browser memory for active work.
[UDS 4] When using a shared workstation, users who are logged in should be able to load and save only their own work.
[UDS 5] When working with an alignment, a user should be offered the option to export their work at any time.
[UDS 6] Upon closing the browser or tab with the alignment editor active, an anonymous user should be reminded to export their work to save a backup in case the browser storage gets cleared.
[UDS 7] Before starting a new alignment, an anonymous user should be prompted to export the current alignment first, if one exists.
[UDS 8] When working with an alignment, an anonymous user should be prompted to login [ a configurable number of ] alignment actions.
[UDS 9] An anonymous user should be able to disable prompts to login.
[UDS 10] When working with an alignment, a logged in user should have a draft copy of their work automatically saved to an alpheios server every [ a configurable number of ] alignment actions.
[UDS 11] When working with an alignment, a logged in user should be offered the option to save a copy of their work to an alpheios server at any time.
[UDS 12] When working with an alignment, a user should be offered the option to publish their alignment data in the Alpheios Online Data Repository at any time. [ Note publishing is different than saving ]
[UDS 13] When publishing alignment data in the Alpheios Online Data Repository, a user must be prompted to first create or login to an Alpheios user account.
[UDS 14] When publishing alignment data in the Alpheios Online Data Repository, a user must be offered the option to make the data public or private.
[UDS 15] When loading the alignment editor in the browser a user should be automatically logged in if the browser has an active, unexpired session.
[UDS 16] When loading the alignment editor in the browser an anoymous user should be asked of they want to start working on their last alignment if one exists in the local store.
[UDS 17] When both a local copy and a remote draft exists, after logging in a user should be offered the option to start working with the last draft saved on the alpheios server, the local copy or neither.
[UDS 18] When only a remote draft exists, after logging in a user should be offered the option to start working with the last draft saved on the alpheios server or not. 
[UDS 19] When only a local copy euxists, after logging in a user should be offered the option to start working with the local copy or not. 
[UDS 20] A logged in user must be offered the ability to remove their published alignments from the Alpheios Online Data Repository.
[UDS 21] A logged in user must be offered the ability to search their published alignments in the Alpheios Online Data Repository.
[UDS 22] A logged in user must be offered the ability to export their published alignments in the Alpheios Online Data Repository.

**Comment by balmas** [comment]
Discussion with @monzug copied from email

> 1) not sure if I understand this one. is it like in the word list (user is logged in and can add words to wordlist on different browsers)? or is it like having a draft copy saved?
> [UDS 1] When workin=g on an alignment, a user should have their work persist locally across browser sessions.

No for the alignment editor, we won't try to keep the local cache and remote copy 100% in sync.  In the wordlist, the remote store is the authority and the local cache is the backup.  This was so that we could support the user accessing Alpheios from anywhere and get consistent data.  The use cases for the alignment editor (for now) don't call for users to be able to access an in progress alignment from multiple applications or browsers.  And the alignment documents might be much larger than a personal wordlist.  So we will prioritize the local cache for saving state and use the remote storage as a backup.

> 2) what's the difference between a draft and a saved copy on alpheios server? regarding the following:
> [UDS 17] When both a local copy and a remote draft exists, after logging in a user should be offered the option to start working with the last draft saved on the alpheios server, the local copy or neither.

See previous answer

> 3) misspelling euxists
> [UDS 19] When only a local copy euxists, after logging in a user should  be offered the option to start working with the local copy or not.
 
Thanks ! (corrected inline above)

> 4) if data is public, then UDS 21 should be valid for all users
> [UDS 21] A logged in user must be offered the ability to search their published alignments in the Alpheios Online Data Repository. 
> - - and all the other public published alignments- -

Yes, thanks! (corrected inline above)

> 5) what do you mean by export?
> [UDS 22] A logged in user must be offered the ability to export their published alignments in the Alpheios Online Data Repository.

Export requirements are described in
https://github.com/alpheios-project/alignment-editor-new/issues/69
https://github.com/alpheios-project/alignment-editor-new/issues/70
https://github.com/alpheios-project/alignment-editor-new/issues/71

**Comment by irina060981** [comment]
I read all the stories and a discussion with Monica - and I see the following separation of things:

- we would have two modes - user is anonymous and user is logged in
- only logged in users could have any data in the remote source
- each user would have local copy of one current alignment (only one or some saved alignments?), (only one for anonymous and several for logged in) ?
- user (anonymous or not) could use only local data (and local tokenizer ?) in offline mode
- there are two remote sources for logged in users - DB and Online Data Repository
- from DB - only owner could get the data
- from Online Repository - data could be retrieved upon access rules (getting data by the link is available only from Online Repository - https://github.com/alpheios-project/alignment-editor-new/issues/84 ?)
- autosaving is done both to local and remote (in parallel?)
- we should have additional block with available alignments (local and both remote sources ?)
- a user could set access rules of the alignment only when it is saved (or anytime from the block with all available alignments ?)

>[UDS 3] When working on an large alignment with many translations, a user should be able to load only a portion of the data into browser memory for active work.

I think this one should be described in more details inside another issue - I don't understand how it should work for now (to be useful for users)

>[UDS 21] A logged in user must be offered the ability to search their published alignments (and all the other public published alignments in the Alpheios Online Data Repository.

And I think this would worth a separate issue with some design details

I like the overall design and I think it would useful for the users to work with alignments this way :)

**Comment by balmas** [comment]
>each user would have local copy of one current alignment (only one or some saved alignments?), (only one for anonymous and several for logged in) ?

They could have multiple copies stored locally, but only one active in the alignment editor at a time.  That's true for the logged in user too -- only one alignment active in the editor, but multiple saved.

**Comment by balmas** [comment]
>user (anonymous or not) could use only local data (and local tokenizer ?) in offline mode

Yes.  It's a good question about the tokenizer. I think we cannot support full functionality for tokenization since that requires the remote service, but we could offer the basic tokenizer.  Or we could just not allow creating new alignments from text while offline.

**Comment by balmas** [comment]
>autosaving is done both to local and remote (in parallel?)

I don't think we should try to keep local and remote in complete sync. So I think these should probably save on different schedules.

**Comment by balmas** [comment]
>we should have additional block with available alignments (local and both remote sources ?)

Definitely for logged in users. Not sure what we want to offer here, if anything for anonymous users. We could allow them to specify a default folder in which they store their local alignments and then they can just use their upload file browser to see what's available.

**Comment by balmas** [comment]
>a user could set access rules of the alignment only when it is saved (or anytime from the block with all available alignments ?)

This only applies to "published" alignments, and I think they should be able to set it from a view of the list of their published alignments.

**Comment by balmas** [comment]
>>[UDS 3] When working on an large alignment with many translations, a user should be able to load only a portion of the data into browser memory for active work.

>I think this one should be described in more details inside another issue - I don't understand how it should work for now (to be useful for users)

>>[UDS 21] A logged in user must be offered the ability to search their published alignments (and all the other public published alignments in the Alpheios Online Data Repository.

>And I think this would worth a separate issue with some design details

Agree on both of these points. They are complex requirements that need further specification.
